libbeat/processors/add_cloud_metadata/providers.go

```diff
@@ -0,0 +1,170 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package add_cloud_metadata
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type provider struct {
+	// Name contains a long name of provider and service metadata is fetched from.
+	Name string
+
+	// Local Set to true if local IP is accessed only
+	Local bool
+
+	// Create returns an actual metadataFetcher
+	Create func(string, *common.Config) (metadataFetcher, error)
+}
+
+type metadataFetcher interface {
+	fetchMetadata(context.Context, http.Client) result
+}
+
+// result is the result of a query for a specific hosting provider's metadata.
+type result struct {
+	provider string        // Hosting provider type.
+	err      error         // Error that occurred while fetching (if any).
+	metadata common.MapStr // A specific subset of the metadata received from the hosting provider.
+}
+
+var cloudMetaProviders = map[string]provider{
+	"alibaba":      alibabaCloudMetadataFetcher,
+	"ecs":          alibabaCloudMetadataFetcher,
+	"azure":        azureVMMetadataFetcher,
+	"digitalocean": doMetadataFetcher,
+	"aws":          ec2MetadataFetcher,
+	"ec2":          ec2MetadataFetcher,
+	"gcp":          gceMetadataFetcher,
+	"openstack":    openstackNovaMetadataFetcher,
+	"nova":         openstackNovaMetadataFetcher,
+	"qcloud":       qcloudMetadataFetcher,
+	"tencent":      qcloudMetadataFetcher,
+}
+
+func selectProviders(list providerList, providers map[string]provider) map[string]provider {
+	return filterMetaProviders(providersFilter(list, providers), providers)
+}
+
+func providersFilter(list providerList, providers map[string]provider) func(string) bool {
+	if len(list) == 0 {
+		return func(name string) bool {
+			ff, ok := providers[name]
+			return ok && ff.Local
+		}
+	}
+	return func(name string) (ok bool) {
+		if ok = list.Has(name); ok {
+			_, ok = providers[name]
+		}
+		return ok
+	}
+}
+
+func filterMetaProviders(filter func(string) bool, fetchers map[string]provider) map[string]provider {
+	out := map[string]provider{}
+	for name, ff := range fetchers {
+		if filter(name) {
+			out[name] = ff
+		}
+	}
+	return out
+}
+
+func setupFetchers(providers map[string]provider, c *common.Config) ([]metadataFetcher, error) {
+	mf := make([]metadataFetcher, 0, len(providers))
+	visited := map[string]bool{}
+
+	for name, ff := range providers {
+		if visited[ff.Name] {
```

Comment:
Is "visited" to account for the redundant short names like "alibaba" and "ecs"? A clarifying comment would be nice

Reply:
Yes.

Reply:
I did add a comment to the loop.